dali/kernels/imgproc/color_manipulation/color_space_conversion_kernel.cuh

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_COLOR_MANIPULATION_COLOR_SPACE_CONVERSION_KERNEL_CUH_
+#define DALI_KERNELS_IMGPROC_COLOR_MANIPULATION_COLOR_SPACE_CONVERSION_KERNEL_CUH_
+
+#include <cuda_runtime_api.h>
+#include <utility>
+#include "dali/kernels/imgproc/color_manipulation/color_space_conversion_impl.h"
+
+namespace dali {
+namespace kernels {
+namespace color {
+
+template <typename Out, typename In>
+DALI_HOST_DEV struct RGB_to_BGR_Converter {
+  static constexpr int out_pixel_sz = 3;
+  static constexpr int in_pixel_sz = 3;
+  static DALI_HOST_DEV vec<out_pixel_sz, Out> convert(vec<in_pixel_sz, In> rgb) {
+    vec<out_pixel_sz, Out> out;
+    out[0] = ConvertSatNorm<Out>(rgb[2]);
+    out[1] = ConvertSatNorm<Out>(rgb[1]);
+    out[2] = ConvertSatNorm<Out>(rgb[0]);
+    return out;
+  }
+};
+
+template <typename Out, typename In>
+DALI_HOST_DEV struct RGB_to_Gray_Converter {
+  static constexpr int out_pixel_sz = 1;
+  static constexpr int in_pixel_sz = 3;
+  static DALI_HOST_DEV vec<out_pixel_sz, Out> convert(vec<in_pixel_sz, In> rgb) {
+    return rgb_to_gray<Out>(rgb);
+  }
+};
+
+template <typename Out, typename In>
+DALI_HOST_DEV struct BGR_to_Gray_Converter {
+  static constexpr int out_pixel_sz = 1;
+  static constexpr int in_pixel_sz = 3;
+  static DALI_HOST_DEV vec<out_pixel_sz, Out> convert(vec<in_pixel_sz, In> bgr) {
+    return RGB_to_Gray_Converter<Out, In>::convert(
+      RGB_to_BGR_Converter<In, In>::convert(bgr));
+  }
+};
+
+template <typename Out, typename In>
+DALI_HOST_DEV struct Gray_to_RGB_Converter {
+  static constexpr int out_pixel_sz = 3;
+  static constexpr int in_pixel_sz = 1;
+  static DALI_HOST_DEV vec<out_pixel_sz, Out> convert(vec<in_pixel_sz, In> gray) {
+    return vec<out_pixel_sz, Out>(gray[0]);
+  }
+};
+
+template <typename Out, typename In>
+DALI_HOST_DEV struct Gray_to_YCbCr_Converter {
+  static constexpr int out_pixel_sz = 3;
+  static constexpr int in_pixel_sz = 1;
+  static DALI_HOST_DEV vec<out_pixel_sz, Out> convert(vec<in_pixel_sz, In> gray) {
+    vec<out_pixel_sz, Out> out;
+    out[0] = ConvertSatNorm<Out>(gray[0]);
```

Comment:
Shouldn't you compress the dynamic range to `itu_r_bt_601` Y  here?

Reply:
Yes, good point.